zebra/zebra_dplane.c

```diff
@@ -885,39 +890,40 @@ void dplane_ctx_fini(struct zebra_dplane_ctx **pctx)
 	dplane_ctx_free(pctx);
 }
 
+/* Init a list of contexts */
+void dplane_ctx_q_init(struct dplane_ctx_list_head *q)
+{
+	dplane_ctx_list_init(q);
+}
+
 /* Enqueue a context block */
-void dplane_ctx_enqueue_tail(struct dplane_ctx_q *q,
+void dplane_ctx_enqueue_tail(struct dplane_ctx_list_head *list,
 			     const struct zebra_dplane_ctx *ctx)
 {
-	TAILQ_INSERT_TAIL(q, (struct zebra_dplane_ctx *)ctx, zd_q_entries);
+	dplane_ctx_list_add_tail(list, (struct zebra_dplane_ctx *)ctx);
 }
 
 /* Append a list of context blocks to another list */
-void dplane_ctx_list_append(struct dplane_ctx_q *to_list,
-			    struct dplane_ctx_q *from_list)
+void dplane_ctx_list_append(struct dplane_ctx_list_head *to_list,
+			    struct dplane_ctx_list_head *from_list)
 {
-	if (TAILQ_FIRST(from_list)) {
```

Comment:
**Note**: the type safe containers don't have a equivalent to `TAILQ_CONCAT`.

The following code snippets execute different work amount:

```c
		if (TAILQ_FIRST(&work_list))
			TAILQ_CONCAT(&(prov->dp_ctx_in_q), &work_list,
				     zd_q_entries);
```

vs

```c
		while ((ctx = dplane_ctx_list_pop(&work_list)) != NULL)
			dplane_ctx_list_add_tail(&(prov->dp_ctx_in_list), ctx);
```

The `TAILQ_CONCAT` implementation is much simpler:

```c
#define TAILQ_CONCAT(head1, head2, field)                                      \
	do {                                                                   \
		if (!TAILQ_EMPTY(head2)) {                                     \
			*(head1)->tqh_last = (head2)->tqh_first;               \
			(head2)->tqh_first->field.tqe_prev =                   \
				(head1)->tqh_last;                             \
			(head1)->tqh_last = (head2)->tqh_last;                 \
			TAILQ_INIT((head2));                                   \
		}                                                              \
	} while (0)
```

I don't think it is necessary to change this here, but if we eventually find it is too much more work then we know where to look at.

Reply:
yes, quite - but as you say, there isn't a "concat" equivalent currently. wouldn't be hard, and probably would be useful...